### ISSUE TITLE
Add android emulator loopback URL to viaduct component

### ISSUE THREAD

**Issue by jrbenny35** [mozilla/application-services#5214]
The android emulator used the URL [10.0.2.2](https://developer.android.com/studio/run/emulator-networking.html) to access the network of the host machine. Currently, trying to send a request in Fenix running in an emulator to a locally hosted URL fails with a TLS error. 

The exact case is hosting a local remote settings, and changing the NIMBUS_ENDPOINT URL before building Fenix. After it is built and the Nimbus client tries to load experiments from the new URL, the request fails with a TLS error.

```
2022-10-27 11:38:58.962 4157-4215/org.mozilla.fenix.debug E/App: Nimbus error: Error fetching experiments from endpoint
    org.mozilla.experiments.nimbus.internal.NimbusException$RequestException: Error sending request: [no-sentry] Validation error: URL does not use TLS protocol.
        at org.mozilla.experiments.nimbus.internal.FfiConverterTypeNimbusError.read(nimbus.kt:1434)
        at org.mozilla.experiments.nimbus.internal.FfiConverterTypeNimbusError.read(nimbus.kt:1419)
        at org.mozilla.experiments.nimbus.internal.FfiConverter$DefaultImpls.liftFromRustBuffer(nimbus.kt:152)
        at org.mozilla.experiments.nimbus.internal.FfiConverterRustBuffer$DefaultImpls.liftFromRustBuffer(nimbus.kt:164)
        at org.mozilla.experiments.nimbus.internal.FfiConverterTypeNimbusError.liftFromRustBuffer(nimbus.kt:1419)
        at org.mozilla.experiments.nimbus.internal.FfiConverterTypeNimbusError.liftFromRustBuffer(nimbus.kt:1419)
        at org.mozilla.experiments.nimbus.internal.FfiConverterRustBuffer$DefaultImpls.lift(nimbus.kt:165)
        at org.mozilla.experiments.nimbus.internal.FfiConverterTypeNimbusError.lift(nimbus.kt:1419)
        at org.mozilla.experiments.nimbus.internal.FfiConverterTypeNimbusError.lift(nimbus.kt:1419)
        at org.mozilla.experiments.nimbus.internal.NimbusException$ErrorHandler.lift(nimbus.kt:1415)
        at org.mozilla.experiments.nimbus.internal.NimbusException$ErrorHandler.lift(nimbus.kt:1414)
        at org.mozilla.experiments.nimbus.internal.NimbusClient.fetchExperiments(nimbus.kt:2130)
        at org.mozilla.experiments.nimbus.Nimbus$fetchExperimentsOnThisThread$1.invoke(Nimbus.kt:432)
        at org.mozilla.experiments.nimbus.Nimbus$fetchExperimentsOnThisThread$1.invoke(Nimbus.kt:430)
        at org.mozilla.experiments.nimbus.Nimbus.withCatchAll(Nimbus.kt:397)
        at org.mozilla.experiments.nimbus.Nimbus.fetchExperimentsOnThisThread$nimbus_release(Nimbus.kt:430)
        at org.mozilla.experiments.nimbus.Nimbus$fetchExperiments$1.invokeSuspend(Nimbus.kt:424)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:923)
```

This would help the testability of Fenix (Firefox for iOS doesn't have this issue) for the Nimbus team specifically.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/SYNC-3461)

**Comment by mhammond** [comment]
To be clear, you'd like the check [here](https://github.com/mozilla/application-services/blob/main/components/viaduct/src/backend.rs#L49) to also allow `10.0.2.2`? At first glance that seems reasonable to me behind a `#[cfg(target_os = "android")]` check, but best I can tell, nothing would stop a real device being on a network where this address was reachable as a normal IP and not a synonym for localhost - so it would probably be ideal if you could get signoff from someone on the Android team that this change would be considered "safe" by them.

**Comment by jrbenny35** [comment]
> but best I can tell, nothing would stop a real device being on a network where this address was reachable as a normal IP and not a synonym for localhost 

sorry I don't follow you here. Could you explain what you mean a bit more?

**Comment by mhammond** [comment]
> sorry I don't follow you here. Could you explain what you mean a bit more?

By special casing `10.0.2.2`, we would be unconditionally treating that address as always meaning localhost, which means it's relatively safe to allow non-https connections to it. However, best I can tell, it would be possible, depending on the network configuration, for a real device to end up with this address being a real, valid, non-loopback address of some other machine on a LAN, which would mean non-https connections are not actually safe.

**Comment by gbrownmozilla** [comment]
Just to clarify a bit more, let's note that 10.0.2.2 is in the IPv4 private address range: 10.0.2.2 can be used by a device on a private network, like a LAN, but not on the public Internet.

**Comment by mhammond** [comment]
To clarify my stance: I'm not against this change, but also don't want to be the decision maker to say it's OK given there's a (tiny) possibility it refers to a real different machine (in the same way I would not want to be the decision maker to drop the https requirement entirely.) It doesn't seem *that* controversial to me, and so I had a quick look in mozilla-central hoping it had precedent for this, but failed to find one outside of tests.

In desktop, it's somewhat common to put things like this behind a pref - something analogous in this world might be to add a new method to nimbus that arranges for this https check to be globally disabled while the app runs, and calling this when a pref or secret menu or something else happens - but that's alot more work - finding someone to 👍 special casing this address would be much easier :)

**Comment by jrbenny35** [comment]
Thanks for all of the feedback. @mhammond what should be my next steps?

**Comment by mhammond** [comment]
@csadilek, @badboy can I please get your opinion here. The tl;dr of the above is that the original request here is to *unconditionally on Android* special-case the address `10.0.2.2` as allowing non-HTTPS connections in Viaduct - currently that's true for all addresses known to be localhost, which `10.0.2.2` is not. This is to aid in deployment and debugging, but is not necessary for normal users.

Some notes:
* I was hoping to find precedent in mozilla-central - it *does* have special handling for this in tests, but not in non-test code.
* Because this is in viaduct, it would impact *all* networking done by all our Rust components, including Nimbus, Glean, Sync, etc (but only Android, so should not impact the use of Viaduct in m-c)

I see 2 options:
1) The easy one is to just hard-code this addresses where we [make the localhost check](https://github.com/mozilla/application-services/blob/main/components/viaduct/src/backend.rs#L49). The downside of this is that it seems possible that on (say) some private lan, this address will be an actual host and Firefox is configured such that (say) remote settings/nimbus experiments come from this address, so allowing non-HTTPS would be defeating the purpose of having the check in the first place.

2) is to somehow allow the app to tell us that this address is OK for this instance of the app - this could look like either a way to say that all addresses can skip the https validation, or even a way to add a single address to be treated as localhost. This is more work because it then implies Fenix needs to grow some way to selectively enable this feature, and possibly additional steps for QA to perform those steps.

3) Any others I've missed? It might be possible to find a middle ground - eg "(1) but only for Nimbus", but I don't think that's a material difference.

The cowboy in me wants to say that (1) is probably fine, is very easy and I can't see a realistic threat scenario, but that might be my naivety. (2) seems like a fair bit of work which bleeds into the Android team to enable.

Now I've written it down, my cowboy hat isn't fitting as well as it did - does it fit for anyone else? Is anyone willing to say "(1) is fine"? If not, any thoughts on concrete actions for (2)?

**Comment by badboy** [comment]
Glean doesn't use viaduct on Android, so would not be affected either way.
What _are_ the URLs affected by this? Can they be changed at runtime?

Hosting something on a public server server over HTTPS is essentially free these days, so that "target" angle already exists regardless of the private IP being allowed.

If it's compile-time only then it is already possible to point it anywhere. But that also isn't a real attack scenario, so kinda harmless.
If it can be changed at runtime any URL is an attack scenario already anyway.

With that in mind I think option 1, adding the check (for Android builds), is ok.

**Comment by csadilek** [comment]
@mhammond @badboy I think we should have our applications opt in to this for debug builds only. Concretely, how about we add a call to [RustHttpConfig](https://github.com/mozilla/application-services/blob/main/components/viaduct/android/src/main/java/mozilla/appservices/httpconfig/HttpConfig.kt#L31) to allow the Android emulator loopback?

So, e.g., in [FenixApplication](https://github.com/mozilla-mobile/fenix/blob/main/app/src/main/java/org/mozilla/fenix/FenixApplication.kt#L466) we would call:
```kotlin
if (Config.channel.isDebug) {
  RustHttpConfig.allowAndroidEmulatorLoopback()
}
```

With the method docs calling out allowing non-TLS traffic to the loopback address (or with a better method name :)). I wouldn't make the address itself configurable, because that's even more of a footgun, but it also doesn't feel quite right to unconditionally allow non-TLS.

**Comment by mhammond** [comment]
I put a branch up at https://github.com/mhammond/application-services/tree/viaduct-android-localhost with the application-services side of this. It has tests for the Rust code, but the changed Android code hasn't been tested at all (other than via `./gradlew :httpconfig:test`, but there are no meaningful tests already in the repo, so that isn't really a test).

So what probably needs to happen next is for someone to test this out via the change @csadilek mentioned in the comment above. I don't intend doing that (mainly because I'm not currently sure how to build Fenix with a local app-services now that a-c has moved to the new repo, but also because it presumably also requires a change so that Fenix actually attempts to use that url), but if someone tells me it works as expected, I'll make the PR for app-services and ask someone (probably @badboy) for review.

**Comment by mhammond** [comment]
> Glean doesn't use viaduct on Android, so would not be affected either way.

huh - that surprises me - I thought viaduct exists purely because there was some edict that all networking done by Fenix must be via Necko.

> What _are_ the URLs affected by this? Can they be changed at runtime?

Only http://10.0.2.2 currently (because from the Android emulator's POV, that is localhost, but the url crate (correctly) doesn't see it that way). I guess it's possible in some future world it might be handy to have some other test server be used without requiring it to be https - but I'm fine with @csadilek's suggestion of not allowing an arbitrary host.

**Comment by mhammond** [comment]
> some edict that all networking done by Fenix must be via Necko.

oh - I suspect Glean does manage to do that, but doesn't need Viaduct to enable it!

**Comment by badboy** [comment]
> > some edict that all networking done by Fenix must be via Necko.
> 
> oh - I suspect Glean does manage to do that, but doesn't need Viaduct to enable it!

Yup, Glean uses `concept.fetch.Client` from a-c. Networking happens on the Kotlin side, that's why we don't need viaduct on the Rust side. It is in use on Desktop though.

**Comment by badboy** [comment]
> > What _are_ the URLs affected by this? Can they be changed at runtime?
> 
> Only http://10.0.2.2 currently (because from the Android emulator's POV, that is localhost, but the url crate (correctly) doesn't see it that way). I guess it's possible in some future world it might be handy to have some other test server be used without requiring it to be https - but I'm fine with @csadilek's suggestion of not allowing an arbitrary host.

Sorry, I meant more which URLs can be changed currently. I assume that's nimbus' experiments endpoint and the sync endpoint, which for testing purposes can be pointed to staging or dev instances?

**Comment by csadilek** [comment]
@mhammond Thanks, I tested this out locally can can confirm that this will work! I see Nimbus requests coming through from the emulator to a local dummy server after calling the new `allowAndroidEmulatorLoopback`. Without it, I am getting the original TLS validation error, of course. So, this will work, butI had to make the following changes locally:

- I had to change this [comparison](https://github.com/mhammond/application-services/commit/54598b0bac94947283cf206433a87cf84b6f79e5#r91525709) to use `host_str()`, so basically: 
```rust
.map(|url| (*url).host_str() != request.url.host_str())
``` 
as @badboy pointed out.
- I had to comment out https://github.com/mhammond/application-services/blob/54598b0bac94947283cf206433a87cf84b6f79e5/components/support/viaduct-reqwest/src/lib.rs#L15 because I was getting compile errors (undefined symbols `GLOBAL_SETTINGS.read_timeout`, `GLOBAL_SETTINGS.connect_timeout`, `GLOBAL_SETTINGS.follow_redirects`. Not sure if that's just me doing something wrong.

- The A-C wrapper of [RustHttpConfig](https://github.com/mozilla-mobile/firefox-android/blob/main/android-components/components/support/rusthttp/src/main/java/mozilla/components/support/rusthttp/RustHttpConfig.kt) needs the new method added, but that's easy. Happy to take care of it.

Then everything worked. 

Local dependency substitution actually still works the same as before. The only difference is the relative path to A-S from A-C, so I had to change it to in A-C's local.properties:
```
autoPublish.application-services.dir=../../application-services
```

**Comment by csadilek** [comment]
@jrbenny35 There were some assumptions made about your use case here though, so wanted to double-check:
- Were you using a local **debug** build of Fenix against a local Nimbus server for manual testing?
- Do you need this for automated integration testing in CI?

If it's the former and you just need it occasionally (and temporarily), then you could also use `ngrok` to create an HTTPS tunnel to localhost, which really comes down to a single command: e.g., `ngrok http 8080`. It will return a temporary HTTPS URL you can use. (Thanks to @tarikeshaq and @jonalmeida for the tip.)

If you need this frequently, for longer periods of time, or in automation, then the proposed API and solution here is much more convenient, as it will just work without any setup.

**Comment by jrbenny35** [comment]
Hey @csadilek, it would be for both. This will help softvision in the short term, and I have some automated plans in the longer term.

**Comment by csadilek** [comment]
@jrbenny35 ok, thanks for confirming! Then I think we should land this.

**Comment by mhammond** [comment]
Fixed by #5270 🚀
